Speed up _extract_graph_with_inputs_outputs

aorenste opened the pull request:
_extract_graph_with_inputs_outputs() does membership testing on the input nodes but often that collection is a list so the test is O(n).  Ensure it's a set before looping over all the nodes.

This change speeds up the internal repro (D57090987) by about 18%:
before:
```
708.88user 15.86system 12:16.19elapsed 98%CPU (0avgtext+0avgdata 12898628maxresident)k
0inputs+91968outputs (3major+3532970minor)pagefaults 0swaps
```
after:
```
583.39user 15.98system 10:10.11elapsed 98%CPU (0avgtext+0avgdata 12895108maxresident)k
0inputs+87488outputs (4major+3374582minor)pagefaults 0swaps
```


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125937

aorenste committed:
Speed up _extract_graph_with_inputs_outputs

[ghstack-poisoned]

oulgen commented on torch/_functorch/partitioners.py:
```diff
@@ -88,6 +88,9 @@ def _extract_graph_with_inputs_outputs(joint_graph, inputs, outputs):
     env = {}
 
+    # Ensure we can ask about our input nodes quickly.
+    inputs = inputs if isinstance(inputs, set) else set(inputs)
```
since we are iterating over this shortly, and want to have deterministic iteration order, should we instead use `dict.fromkeys` and update whatever is making inputs set in the first place?

aorenste replied:
Lol - I actually meant to put this line after the iteration (and in my internal version I did - I just copied the change over to the OSS version poorly).  But I also realized that it's probably correct to do `node in env` instead.

aorenste committed:
Update on "Speed up _extract_graph_with_inputs_outputs"

_extract_graph_with_inputs_outputs() does membership testing on the input nodes but often that collection is a list so the test is O(n).  Ensure it's a set before looping over all the nodes.

This change speeds up the internal repro (D57090987) by about 18%:
before:
```
708.88user 15.86system 12:16.19elapsed 98%CPU (0avgtext+0avgdata 12898628maxresident)k
0inputs+91968outputs (3major+3532970minor)pagefaults 0swaps
```
after:
```
583.39user 15.98system 10:10.11elapsed 98%CPU (0avgtext+0avgdata 12895108maxresident)k
0inputs+87488outputs (4major+3374582minor)pagefaults 0swaps
```





[ghstack-poisoned]

aorenste committed:
Update on "Speed up _extract_graph_with_inputs_outputs"

_extract_graph_with_inputs_outputs() does membership testing on the input nodes but often that collection is a list so the test is O(n).  Ensure it's a set before looping over all the nodes.

This change speeds up the internal repro (D57090987) by about 18%:
before:
```
708.88user 15.86system 12:16.19elapsed 98%CPU (0avgtext+0avgdata 12898628maxresident)k
0inputs+91968outputs (3major+3532970minor)pagefaults 0swaps
```
after:
```
583.39user 15.98system 10:10.11elapsed 98%CPU (0avgtext+0avgdata 12895108maxresident)k
0inputs+87488outputs (4major+3374582minor)pagefaults 0swaps
```





[ghstack-poisoned]

anijain2305 reviewed (approved):
Awesome!

aorenste commented:
@pytorchbot merge

henryhu6 commented:
Nice!

I think it this compilation with cache, right? 

Because cold start was taking >6000sec for me.

aorenste commented:
> Nice!
> 
> I think it this compilation with cache, right?
> 
> Because cold start was taking >6000sec for me.

That's right - that timing is with the cache. I imagine it will help the cold start time less.
